### ISSUE TITLE
Treat anything after UnitTest.fail as unreachable

### ISSUE THREAD

**Issue by nijel** [astral-sh/ruff#23509]
### Summary

Pylint introduced this recently (see https://github.com/pylint-dev/pylint/issues/10743) and this makes linting outcome different for several rules between pylint and ruff.

For me, the specific rule I'm hitting is RET503:

```python
from unittest import TestCase
import os

class TestX(TestCase):
  def assert_foo(self) -> str:
    if 'FOO' in os.environ:
        return ""
    self.fail()
```

Adding code after UnitTest.fail is (correctly) detected by pylint as unreachable code, while ruff insists of adding `return ""` there.

**Comment by ntBre** [comment]
I think for general reachability analysis, we should probably wait to borrow it from ty (https://github.com/astral-sh/ruff/issues/5474 is somewhat related and labeled https://github.com/astral-sh/ruff/labels/type-inference). However, for `RET503` in particular, I think we could add `unittest.TestCase.fail` to the known list of `NoReturn` functions first added in https://github.com/astral-sh/ruff/pull/2701.

**Comment by spider-yamet** [comment]
@ntBre @MichaReiser Currently I’m thinking small fix for RET503. As you mentioned earlier, just add one more narrow case for unittest: if the call is .fail(...) and inside a method of class that inherits from unittest.TestCase , then treat that call for terminal as well.

If you don't have any other opinion, I'd like to open PR. Would appreciate your feedback before opening, Thanks.

**Comment by spider-yamet** [comment]
Any opinion from your side? Waiting for discussion. :)
